common/src/main/java/org/cbioportal/cmo/pipelines/common/util/HttpClientWithTimeoutAndRetry.java

```diff
@@ -0,0 +1,173 @@
+    }
+
+    public int getNumberOfRequestsAttempted() {
+        return numberOfRequestsAttempted;
+    }
+
+    public int getLastTimeoutUsed() {
+        return lastTimeoutUsed;
+    }
+
+    public boolean getReachedDropDeadInstant() {
+        return reachedDropDeadInstant;
+    }
+
+    public RestClientException getLastRestClientException() {
+        return lastRestClientException;
+    }
+
+    public String getLastResponseBodyStringAfterException() {
+        return lastResponseBodyStringAfterException;
+    }
+
+    public boolean exceptionCausedByTimeout(RestClientException e) {
+        return e.getRootCause() instanceof SocketTimeoutException && e.getRootCause().getMessage().toLowerCase().contains("timed out");
+    }
+    /* the exchange method takes similar arguments to HttpTemplate.exchange and it sets the intial timeout from data member initialTimeout.
+     * if the timeout is reached (either for forming a connection or waiting to receive a response) a RestClientException is raised. If
+     * the exception is due to timeout, the timeout is adjusted according to the HttpRequestTimeoutProgression object, and the request
+     * is reattempted. A final DropDeadInstant is also monitored and if it arrives a null response is returned.
+    */
+    public <T> ResponseEntity<T> exchange(String url, HttpMethod method, HttpEntity requestEntity, Map<String,?> uriVariables, Class<T> returnType) {
+		HttpRequestTimeoutProgression timeoutProgression = new HttpRequestTimeoutProgression(initialTimeout, maximumTimeout, dropDeadInstant);
+		SimpleClientHttpRequestFactory requestFactory = new SimpleClientHttpRequestFactory();
+        RestTemplate restTemplate = new RestTemplate();
+		restTemplate.setRequestFactory(requestFactory);
+        if (uriVariables == null) {
+            uriVariables = new HashMap<String, String>();
+        }
+        numberOfRequestsAttempted = 0;
+        lastTimeoutUsed = 0;
+        reachedDropDeadInstant = false;
+        lastRestClientException = null;
+        lastResponseBodyStringAfterException = null;
+        while (Instant.now().isBefore(dropDeadInstant)) {
+            try {
+                lastRestClientException = null; // reset before each request
+                lastResponseBodyStringAfterException = null; // reset before each request
+                numberOfRequestsAttempted = numberOfRequestsAttempted + 1;
+                lastTimeoutUsed = timeoutProgression.getNextTimeoutForRequest();
+		        requestFactory.setConnectTimeout(lastTimeoutUsed);
+		        requestFactory.setReadTimeout(lastTimeoutUsed);
+                ResponseEntity<T> response = restTemplate.exchange(url, method, requestEntity, returnType, uriVariables);
+                if (response.getStatusCode().is5xxServerError()) {
+                    pauseForMilliseconds(lastTimeoutUsed);
+System.out.println("got error: " + response.getStatusCode().toString());
+                    pauseForMilliseconds(lastTimeoutUsed);
```

Comment:
Did you intend for this additional pause?

Reply:
This was testing code accidentally left in. Now removed.